engine/table/src/main/java/io/deephaven/engine/table/impl/multijoin/StaticMultiJoinStateManagerTypedBase.java

```diff
@@ -0,0 +1,218 @@
+
+    protected StaticMultiJoinStateManagerTypedBase(ColumnSource<?>[] tableKeySources,
+            ColumnSource<?>[] keySourcesForErrorMessages,
+            int tableSize,
+            double maximumLoadFactor) {
+        this.keySourcesForErrorMessages = keySourcesForErrorMessages;
+
+        this.tableSize = tableSize;
+        Require.leq(tableSize, "tableSize", MAX_TABLE_SIZE);
+        Require.gtZero(tableSize, "tableSize");
+        Require.eq(Integer.bitCount(tableSize), "Integer.bitCount(tableSize)", 1);
+        Require.inRange(maximumLoadFactor, 0.0, 0.95, "maximumLoadFactor");
+
+        mainKeySources = new WritableColumnSource[tableKeySources.length];
+        chunkTypes = new ChunkType[tableKeySources.length];
+
+        outputKeySources = new WritableColumnSource[tableKeySources.length];
+
+        for (int ii = 0; ii < tableKeySources.length; ++ii) {
+            chunkTypes[ii] = tableKeySources[ii].getChunkType();
+            mainKeySources[ii] = InMemoryColumnSource.getImmutableMemoryColumnSource(tableSize,
+                    tableKeySources[ii].getType(), tableKeySources[ii].getComponentType());
+            outputKeySources[ii] = ArrayBackedColumnSource.getMemoryColumnSource(tableSize,
+                    tableKeySources[ii].getType(), tableKeySources[ii].getComponentType());
+        }
+
+        this.maximumLoadFactor = maximumLoadFactor;
+
+        ensureCapacity(tableSize);
+    }
+
+    private void ensureCapacity(int tableSize) {
+        slotToOutputRow.ensureCapacity(tableSize);
+        for (WritableColumnSource<?> mainKeySource : mainKeySources) {
+            mainKeySource.ensureCapacity(tableSize);
+        }
+    }
+
+    BuildContext makeBuildContext(ColumnSource<?>[] buildSources, long maxSize) {
+        return new BuildContext(buildSources, (int) Math.min(CHUNK_SIZE, maxSize));
+    }
+
+    @Override
+    public void build(final Table table, ColumnSource<?>[] keySources, int tableNumber) {
+        if (table.isEmpty()) {
+            return;
+        }
+        final LongArraySource tableRedirSource = indexSources.get(tableNumber);
+        try (final BuildContext bc = makeBuildContext(keySources, table.size())) {
+            buildTable(bc, table.getRowSet(), keySources, new LeftBuildHandler(tableRedirSource, tableNumber));
+        }
+    }
+
+    private class LeftBuildHandler implements TypedHasherUtil.BuildHandler {
+        final LongArraySource tableRedirSource;
+        final long tableNumber;
+
+        private LeftBuildHandler(LongArraySource tableRedirSource, long tableNumber) {
+            this.tableRedirSource = tableRedirSource;
+            this.tableNumber = tableNumber;
+        }
+
+        @Override
+        public void doBuild(RowSequence chunkOk, Chunk<Values>[] sourceKeyChunks) {
+            final long maxSize = numEntries + chunkOk.intSize();
+            slotToOutputRow.ensureCapacity(maxSize);
```

Comment:
I find it confusing that this is happening here, and in `ensureCapacity`. I think it's either like `outputKeySources` *or* like `mainKeySources`. Which is correct?

Reply:
This is incorrect here, has been removed.